### ISSUE TITLE
The execution never ends

### ISSUE THREAD

**Issue by emilioriosvz** [substack/tape#216]
When I run the test, the execution never ends and what I have to do is in the last test put a `process.exit(1)`

My `package.json` looks like this: 

```
"scripts": {
    "test": "tape ./test/*.js | tap-spec"
  }
```

there a better way?

**Comment by taylorhakes** [comment]
The tests should finish automatically. Did you forget to put `t.end()` in one of your tests?

**Comment by growlove** [comment]
I'm having this same issue. It doesn't matter whether I use t.end() or t.plan(). After running all the tests, the process sits until I manually kill it.

**Comment by substack** [comment]
Tests are ordinary node processes, so if you have open handles, the process will hang as it ought to.

**Comment by emilioriosvz** [comment]
that's true, I open connections to databases and then compare answers (among other things)
I guess that is why the process does not finish.
I have resolved to create a file named `ztest.js`:
```
const tape = require('tape')

tape('SUMMARY', function (t) {
  process.exit(0)
})

```
The file begins with z to be the last to be called, and and as I said before, this is my `package.json`:

```
"scripts": {
    "test": "tape ./test/*.js | tap-spec"
}
```

**Comment by tcurdt** [comment]
Just run into the same thing. Also open connections in a db connection pool.

Now the problem is - I am not sure how to gracefully shut everything down. Given that `supertest` is calling the `express` app directly I have no clue if there is a way to hook into this.

The interesting bit - it just works with `mocha`.

    var request = require('supertest');

    describe('bla', function(){
      it('test', function(done) {

        request(app)
          .post('/auth/token')
          .send({
            username: 'the_username',
            password: 'the_password'
          })
          .expect(200)
          .expect('Content-Type', /json/)
          .end(function(err, res) {
            done();
          });

      });
    });

    test('test', function(t) {

      request(app)
        .post('/auth/token')
        .send({
          username: 'the_username',
          password: 'the_password'
        })
        .expect(200)
        .expect('Content-Type', /json/)
        .end(function(err, res) {
          t.end();
        });

    });

**Comment by wavded** [comment]
Tape doesn't do any magic like mocha, if anything keeps the event loop open, it will stay open and you are responsible for closing any connections, etc.  I personally think this method helps me understand my application better and avoids issues I've had with mocha's magic getting in the way.

That said, it can be confusing to track down what actually is keeping the event loop open!  In the case of express apps, usually its that I have a database connection open (for session storage for example) and need to either `unref` that in my tests or close it down at the end of my tests.

**Comment by Raynos** [comment]
There is a module called `leaked-handles` that can be used for debugging.

**Comment by tcurdt** [comment]
https://github.com/Raynos/leaked-handles nice, thanks

**Comment by recursivefunk** [comment]
Boom @Raynos @tcurdt thanks for posting - needed this.

**Comment by jlertle** [comment]
This may be of interest: http://puigcerber.com/2015/11/27/testing-express-apis-with-tape-and-supertest/

**Comment by chrisfrancis27** [comment]
Just encountered this and fixed it by stopping the HTTP server (hapi) and disconnecting from the Postgres database (knex) in my teardown test.

**Comment by Attrash-Islam** [comment]
I've closed the db connection and it well perfect.

**Comment by 5h1rU** [comment]
This can be hepful

```js

// ES6
test.onFinish(() => process.exit(0));

// ES5
test.onFinish(function() { 
  process.exit(0)
});

**Comment by SteveALee** [comment]
I found when running in browser-run on windows my tests nearly always hung on the command line and even ctrl c had no effect. The solution was to throw at the end of the test

**Comment by shaunwarman** [comment]
Quite the rabbit hole. Had an open handle to mongoDB using mongoose. Just needed to close the connection.

**Comment by t3hmrman** [comment]
Just ran into this as well, want to note to anyone that this happens if you use `setInterval` in your tests as well (not just things like open db connections)

**Comment by camsjams** [comment]
For some people here, make sure that when in test mode (usually best found via `NODE_ENV`) your server doesn't listen for any connections.

**Comment by ORESoftware** [comment]
@substack tests are ordinary node.js processes, and therefore a test runner should force-close (with `process.exit(x)`) the test process after the final test case?

**Comment by substack** [comment]
Calling `process.exit()` in your tests is not the best way to solve this problem, since a callback might otherwise double-fire and you wouldn't know about it. It's also useful to know if your code is leaking handles. Here's a way you can set up your tests to avoid this problem:

``` js
var test = require('tape')

var db
test('setup', function (t) {
  db = openDatabase()
})

test.onFinish(function () {
  db.close()
})
```

**Comment by ORESoftware** [comment]
well process.exit is also bad because you don't know what exit code to use, the test runner knows the exit code.

**Comment by mattlub** [comment]
@substack thanks for your answer, changing to your structure works for me. The code below was what I had before and it was not working- the db connection was opened but the test was never running, can you tell why?
```js
var test = require('tape')

var db = openDatabase()
db.on('open', function () {

  test('setup', function (t) {
    t.ok('test passing')
    t.end()
  })

  test.onFinish(function () {
    db.close()
  })
})
```

**Comment by t3hmrman** [comment]
Hey @mattlub I think I can answer that actually:

```
var test = require('tape');
var DB_CONN;

test('setup database', {timeout: 10000}, function(t) {
  openDatabase()
    .on('open'), function() { 
      t.comment('Database connection established');
      t.end();
    })
})

... tests that use the database ...

test('Close the database connection', function(t) {
// not sure if this takes a callback or produces or promise, but I went ahead and assumed it took a callback -- I assume DB.close() is not synchronous?
  DB.close(function() { 
    t.comment('Database connection closed');
    t.end();
  }); 
})
```

**Comment by des-des** [comment]
@mattlub @t3hmrman t.plan is also seems to fix

**Comment by jakubrpawlowski** [comment]
Run into this yesterday and yes closing mongoose connection did the job!

**Comment by CleitonDants** [comment]
`const test = require('tape');
const supertest = require('supertest');
const app = require('../app');
const docJson = require('./docJson');
const mongoose = require('mongoose');

function dropDB() {
  mongoose.connection.db.dropCollection('ingr_users');
}

test('Init database:', {timeout: 10000}, (t) => {
  mongoose.connection
    .on('open', () => {
      t.comment('Database connection established');
      dropDB();
      t.end();
    });
});

test('========== POST api/v1/user ==========', (t) => {
  supertest(app)
    .post('/api/v1/user/')
    .send(docJson)
    .expect('Content-Type', /json/)
    .expect(200)
    .end((err, res) => {
      t.error(err, 'No error returned');
      t.assert(res.body.name === docJson.name, 'Name inserted!');
      t.assert(res.body.idade === docJson.idade, 'Age inserted!');
      t.assert(res.body.email === docJson.email, 'Email inserted!');
      t.end();
    });
});

test('========== GET api/v1/user ==========', (t) => {
  supertest(app)
    .get(`/api/v1/user/`)
    .expect('Content-Type', /json/)
    .expect(200)
    .end((err, res) => {
      t.error(err, 'No error returned');
      t.assert(res.body[0].name === docJson.name, 'Name catched!');
      t.assert(res.body[0].idade === docJson.idade, 'Age catched!');
      t.assert(res.body[0].email === docJson.email, 'Email catched!');
      t.end();
    });
});

test('Disconnect database:', (t) => {
  mongoose.connection.close(() => {
    t.comment('Database connection closed');
    t.end();
  });
});

`

Guy, I can't explain how, but now after implement more and more tests, I stop with "npm run test --silence" and I run "npm run test" before sends you the code, magically it's working without the error. I am scared because I don't undestand what happened hahaha.

**Comment by CleitonDants** [comment]
Guy, I can't explain how, but now after implement more and more tests, I stop with "npm run test --silence" and I run "npm run test" before sends you the code, magically it's working without the error. I am scared because I don't undestand what happened hahaha.

**Comment by jorgesnk** [comment]
I solved the problem using

tape.plan ()  

tape.on Finish (() => {process.exit (0)})

**Comment by t3hmrman** [comment]
Just to reiterate, as others have noted here, the right way to solve this is to write code that does *not* leave NodeJS hanging.

You can choose to either use `tape.plan` or just `tape.assert` & call `t.end` yourself, but your tests should not hang  -- they end up hanging whenever you do stuff that hangs the runtime/event loop, like:

- uncaught exceptions
- never ending `setInterval`
- libraries that you run which use the event loop that need to be stopped/cleaned up

Generally, for setup of persistent testing utilities, just make a pseudo-test that sets the utility up. Here's an example from some recent code I've written with `tape` -- I'm not claiming it's the golden way, but it's *a* way:

```typescript
let TEST_DB: DB; // Global DB component for this test file
let TEST_UGC: LocalDirUGC; // Global DB component for this test file
let app: Express = buildApplication();

// Tape test that sets up the DB -- passing a callback to set up the global TEST_DB in this file
TestUtil.setupTestDB((db: DB) => {
    TEST_DB = db;
    return TEST_DB.setup(app);
});

TestUtil.setupTestUGC((ugc: LocalDirUGC) => {
    TEST_UGC = ugc;
    return TEST_UGC.setup(app);
});

test("document creation by ugc upload works", (t: Test) => {
    // ... your test code, which uses TEST_DB
});

// ... other tests which make use of TEST_DB ...

TestUtil.cleanupTestDB(() => TEST_DB);
```
It's perhaps a little hard to see, but `setupTestDB` is a function that when evaluated *runs* the the test -- but by having `setupTestDB` be a function, you can pass it a callback to use to set a variable in the script-global state for the test file. `setupTestDB` finishes when the setup is finished, so the global variable is set before any other tests run.

You can use your global (in my case `TEST_DB`) in subsequent tests since `tape` runs tests in a sequential manner. I normally add some code to the test instance to ensure that I can reset it between tests if necessary (and if creating a new one is costly).

`setupTestDB` looks like this:
```typescript
/**
 * Set up a database for test
 *
 * @param {Function} cb - Callback to run when the test database is set up (i.e. you could set a global value to the DB)
 */
export function setupTestDB(cb?: (db: DB) => Promise<any>) {
    // Run a test that setups the datbase
    test("setting up database for test", (t: Test) => {
        makeTestDB()
            .then((db: DB) => {
                t.ok(db, `db created for testing (database: ${db.getDatabaseName()})`);
                return cb ? cb(db) : Promise.resolve();
            })
            .then(() => t.end())
            .catch(t.end);
    })
}
```
and `cleanupTestDB` looks like this:

```typescript
/**
 * Cleanup a database for test
 *
 * @param {Function} getDB - A function that returns the DB object
 */
export function cleanupTestDB(getDB: () => DB) {
    // Run a test that setups the datbase
    test("cleaning up database used in test(s)", (t: Test) => {
        // Get the DB out of the closure
        const db = getDB();
        if (typeof db === "undefined") { throw new Error("DB is undefined"); }

        const dbName = db.getDatabaseName();

        // Retrieve the default and actual connection being used by the db (in test they will likely be different, one pointing to a temp db)
        Promise.all([
            db.getDefaultConnection(),
            db.getConnection(),
        ])
            .then(([defaultConn, conn]) => {
                // Ensure conn is at least defined
                if (!conn) { return Promise.reject(new Error("DB is disconnected (not even default)")); }

                // Ensure we don't ever try to drop the default (ormconfig.js configured) DB, or a db named "happilease"
                if (defaultConn === conn || dbName === "master") {
                    t.comment("DB name is [master], will not remove");
                    return Promise.resolve();
                }

                // Close the connection then use default to drop the associated DB
                return conn.close()
                    .then(() => defaultConn.query(`DROP DATABASE ${dbName}`));
            })
            .then(() => t.ok(db, `db ${dbName} dropped`))
            .then(() => t.end())
            .catch(t.end);
    })
}
```
(apologies, some of this code is very specific to `typeorm`), but it should give a good example of the pattern I'm employing.

BTW, this pattern also scales well to parallel suite running with [`bogota`](https://www.npmjs.com/package/bogota) -- since bogota runs the suites (test files) in parallel but still runs the *tests inside* serially.

**Comment by t3hmrman** [comment]
Just to reiterate, as others have noted here, the right way to solve this is to write code that does *not* hang the NodeJS runtime. You can choose to either use `tape.plan` & a set number of tests or just `tape.assert` & call `t.end`, but your tests should not hang  -- they end up hanging whenever you do stuff that hangs the runtime/event loop, like:

- uncaught exceptions
- never ending `setInterval`
- libraries that you run which use the event loop that need to be stopped/cleaned up

Generally, for setup of persistent testing resources, you can make a pseudo-test that sets the utility up. Here's an example from some recent code I've written with `tape` -- I'm not claiming it's the golden way, but it's *a* way:

```typescript
let TEST_DB: DB; // Global DB component for this test file
let TEST_UGC: LocalDirUGC; // Global DB component for this test file
let app: Express = buildApplication();

// Tape test that sets up the DB -- passing a callback to set up the global TEST_DB in this file
TestUtil.setupTestDB((db: DB) => {
    TEST_DB = db;
    return TEST_DB.setup(app);
});

TestUtil.setupTestUGC((ugc: LocalDirUGC) => {
    TEST_UGC = ugc;
    return TEST_UGC.setup(app);
});

test("document creation by ugc upload works", (t: Test) => {
    // ... your test code, which uses TEST_DB
});

// ... other tests which make use of TEST_DB ...

TestUtil.cleanupTestDB(() => TEST_DB);
```
It's perhaps a little hard to see, but `setupTestDB` is a function that when evaluated *runs* the the test -- but by having `setupTestDB` be a function, you can pass it a callback to use to set a variable in the script-global state for the test file. `setupTestDB` finishes when the setup is finished, so the global variable is set before any other tests run.

You can use your global (in my case `TEST_DB`) in subsequent tests since `tape` runs tests in a sequential manner. I normally add some code to the test instance to ensure that I can reset it between tests if necessary (and if creating a new one is costly).

`setupTestDB` looks like this:
```typescript
/**
 * Set up a database for test
 *
 * @param {Function} cb - Callback to run when the test database is set up (i.e. you could set a global value to the DB)
 */
export function setupTestDB(cb?: (db: DB) => Promise<any>) {
    // Run a test that setups the datbase
    test("setting up database for test", (t: Test) => {
        makeTestDB()
            .then((db: DB) => {
                t.ok(db, `db created for testing (database: ${db.getDatabaseName()})`);
                return cb ? cb(db) : Promise.resolve();
            })
            .then(() => t.end())
            .catch(t.end);
    })
}
```
and `cleanupTestDB` looks like this:

```typescript
/**
 * Cleanup a database for test
 *
 * @param {Function} getDB - A function that returns the DB object
 */
export function cleanupTestDB(getDB: () => DB) {
    // Run a test that setups the datbase
    test("cleaning up database used in test(s)", (t: Test) => {
        // Get the DB out of the closure
        const db = getDB();
        if (typeof db === "undefined") { throw new Error("DB is undefined"); }

        const dbName = db.getDatabaseName();

        // Retrieve the default and actual connection being used by the db (in test they will likely be different, one pointing to a temp db)
        Promise.all([
            db.getDefaultConnection(),
            db.getConnection(),
        ])
            .then(([defaultConn, conn]) => {
                // Ensure conn is at least defined
                if (!conn) { return Promise.reject(new Error("DB is disconnected (not even default)")); }

                // Ensure we don't ever try to drop the default (ormconfig.js configured) DB, or a db named "master"
                if (defaultConn === conn || dbName === "master") {
                    t.comment("DB name is [master], will not remove");
                    return Promise.resolve();
                }

                // Close the connection then use default to drop the associated DB
                return conn.close()
                    .then(() => defaultConn.query(`DROP DATABASE ${dbName}`));
            })
            .then(() => t.ok(db, `db ${dbName} dropped`))
            .then(() => t.end())
            .catch(t.end);
    })
}
```
(apologies, some of this code is very specific to `typeorm`), but it should give a good example of the pattern I'm employing.

BTW, this pattern also scales well to parallel suite running with [`bogota`](https://www.npmjs.com/package/bogota) -- since bogota runs the suites (test files) in parallel but still runs the *tests inside* serially.

**Comment by yhojann-cl** [comment]
Same problem using promises and sockets, the execution of script is end (debugged) but the test is not end:

```
const test = require('tape');
const { DNSResolve } = require('./');

test('Resolve UDP', async (t) => {
    const dnsResolve = new DNSResolve();
    await dnsResolve.resolve({
        server: 'udp://8.8.8.8:53',
        hostname: 'google.com',
        type: 'A'
    })
    .then(results => {
        results = results[0]
        t.ok(results.header.id == 13330);
        t.ok(results.header.type == 'response');
    })
    .catch(e => {
        t.error(e);
    });

    console.log('It is end!');
    t.end();
})

test.onFinish(() => process.exit(0));
```

The `console.log` is printed on finish of execution but the process is not finished.

**Comment by ljharb** [comment]
@yhojann-cl you don't need the `onFinish` or the `t.end()` there, if you're using tape 5+; and that won't work at all using tape < 5, because you need to use `t.plan` for async tests (tape 5 added Promise support)

**Comment by yhojann-cl** [comment]
But creating an advance test plan is more tedious and more prone to errors when inserting new tests within a large code, it is safer that the quantity is not declared and declare the completion at a certain point, in this way it is not prone to errors due to test recounting.

**Comment by ljharb** [comment]
@yhojann-cl yes, which is why tape v5 added support for a test function returning a Promise - which is why in your example, assuming you're using tape 5, you can omit the `t.end()` and also the `onFinish`, and it should Just Work. If it doesn't, please file a new issue instead of resurrecting this ancient one :-)
